### PR TITLE
INTLY-2757 user level sso

### PR DESCRIPTION
**Pull request by austincunningham**
### Description 
Add launcher operator 
Manifest PR: https://github.com/integr8ly/manifests/pull/59

### Verification steps
- Create the operator source which points to my app registry:
```yml
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  finalizers:
    - finalizer.operatorsources.operators.coreos.com
  name: integreatly-operators
  namespace: openshift-marketplace
spec:
  authorizationToken: {}
  displayName: Integreatly Operators austincunningham
  endpoint: 'https://quay.io/cnr'
  publisher: austincunningham Publisher
  registryNamespace: austincunningham
  type: appregistry
```
- Create a subscription to the Integreatly operator from the Operator Hub, then create an `installation` CR 
```yml
apiVersion: integreatly.org/v1alpha1
kind: Installation
metadata:
  name: example-installation
spec:
  type: workshop
  namespacePrefix: integreatly-
  routingSubdomain: apps.cluster-id.openshiftworkshop.com
  masterUrl: https://console-openshift-console.apps.cluster-id.openshiftworkshop.com
  apiUrl: https://api.cluster-id.openshiftworkshop.com
  selfSignedCerts: true
```

- Verify that the `integreatly-user-sso` namespace is created
- Verify that the `rhsso-operator` pod in the namespace 
- Verify that the user-sso route is accessible 
- Verify that the you can login to the realm user-sso using the sso admin credentials

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,9 +1,10 @@
 ORG=integreatly
 NAMESPACE=integreatly
+RHMDSNAMESPACE=mobile-developer-console
```
Comment (dimitraz) [inline Makefile]: What is this for?

File: pkg/controller/installation/products/rhssouser/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	r.Config.SetHost(string(kcURLBytes))
+	err = r.ConfigManager.WriteConfig(r.Config)
+	if err != nil {
+		return pkgerr.Wrap(err, "could not update keycloak config for user-sso")
+	}
+	return nil
+}
+
+func (r *Reconciler) setupOpenshiftIDP(ctx context.Context, kcr *aerogearv1.KeycloakRealm, serverClient pkgclient.Client) error {
+	_, err := r.oauthv1Client.OAuthClients().Get(oauthId, metav1.GetOptions{})
```
Comment (dimitraz) [inline pkg/controller/installation/products/rhssouser/reconciler.go]: You should be able to use `r.ReconcileOauthClient` now 

https://github.com/integr8ly/integreatly-operator/pull/91/files#diff-329c149b9583e9a4fcdb670a2ba866f0R275-R289

File: README.md [hunk 3]
```diff
@@ -67,6 +67,11 @@ oc create -f https://raw.githubusercontent.com/integr8ly/integreatly-operator/ma
 oc create -f https://raw.githubusercontent.com/integr8ly/integreatly-operator/master/deploy/role_binding.yaml
 ```
 
+Create a `Role` and `RoleBinding` for RHMDS
```
Comment (sedroche) [inline README.md]: Shouldn't this be created from a manifest?

File: Makefile (hunk 1 above)
Comment (austincunningham) [reply Makefile]: its for mdc https://github.com/integr8ly/installation/blob/0dccb5958f2b746532b767a2d4da2d5e3ca054a7/roles/rhsso-user/tasks/main.yml#L22 doens't actually do anything at the moment

File: README.md (hunk 3 above)
Comment (austincunningham) [reply README.md]: role is applied to the RHMCS namespace which doesn't do anything at the moment.

Comment (austincunningham) [reply README.md]: but maybe it should ?

File: Makefile (hunk 1 above)
Comment (sedroche) [reply Makefile]: If its not part of user rhsso lets leave it out

Comment (austincunningham) [reply Makefile]: has a chat with John, removing mdc stuff

Comment (dimitraz) [reply Makefile]: :+1:

File: pkg/controller/installation/products/rhssouser/reconciler.go (hunk 2 above)
Comment (dimitraz) [reply pkg/controller/installation/products/rhssouser/reconciler.go]: You can also use `r.ReconcileFinalizer` (just merged) to clean up oauth clients and any other cluster level resources created by your reocnciler

File: pkg/controller/installation/products/rhssouser/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+		kcr.Spec.ID = keycloakRealmName
+		kcr.Spec.Realm = keycloakRealmName
+		kcr.Spec.DisplayName = keycloakRealmName
+		kcr.Spec.Enabled = true
+		kcr.Spec.EventsListeners = []string{
+			"metrics-listener",
+		}
+
+		if kcr.Spec.Users == nil {
+			kcr.Spec.Users = []*aerogearv1.KeycloakUser{CustomerAdminUser}
```
Comment (sedroche) [inline pkg/controller/installation/products/rhssouser/reconciler.go]: There is a ticket to remove customer-admin user so we can remove that from here:
https://issues.jboss.org/browse/INTLY-2932

File: pkg/controller/installation/products/rhssouser/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+		}
+		users, err := syncronizeWithOpenshiftUsers(kcr.Spec.Users, ctx, serverClient)
```
Comment (sedroche) [inline pkg/controller/installation/products/rhssouser/reconciler.go]: @maleck13 This code syncs the openshift users into the user rhsso. Do we want that?

File: pkg/controller/installation/products/rhssouser/reconciler.go [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+func (r *Reconciler) setupOpenshiftIDP(ctx context.Context, inst *v1alpha1.Installation, kcr *aerogearv1.KeycloakRealm, serverClient pkgclient.Client) error {
+	oauthc := &oauthv1.OAuthClient{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: oauthId,
+		},
+		Secret: clientSecret,
+		RedirectURIs: []string{
+			r.Config.GetHost() + "/auth/realms/openshift/broker/openshift-v4/endpoint",
```
Comment (dimitraz) [inline pkg/controller/installation/products/rhssouser/reconciler.go]: Should `openshift` be changed to the new realm name?

Comment (austincunningham) [reply pkg/controller/installation/products/rhssouser/reconciler.go]: :+1: redirect uri is pointing to https://sso-integreatly-user-sso.apps.aucunnin.cloudservices.rhmw.io/auth/realms/user-sso/broker/openshift-v4/endpoint

File: pkg/controller/installation/products/rhssouser/reconciler.go (hunk 5 above)
Comment (sedroche) [reply pkg/controller/installation/products/rhssouser/reconciler.go]: @austincunningham  Checked with Phil and this is not needed

Comment (austincunningham) [reply pkg/controller/installation/products/rhssouser/reconciler.go]: @sedroche removed it and it appears to work fine.

**Pull request merged**